keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,240 @@
+authors: all of us
+creation-date: 2020-09-01
+status: draft
+---
+```
+
+# Standardize ExternalSecret CRD
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Terminology](#terminology)
+- [Use-Cases](#use-cases)
+- [Proposal](#proposal)
+  - [API](#api)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This is a proposal to standardize the `ExternalSecret` CRD in an combined effort through all projects that deal with syncing external secrets. This proposal aims to do find the _common denominator_ for all users of `ExternalSecrets`.
+
+## Motivation
+
+There are a lot of different projects in the wild that essentially do the same thing: sync secrets with Kubernetes. The idea is to unify efforts into a single project that serves the needs of all users in this problem space.
+
+As a starting point i would like to define a **common denominator** for a CustomResourceDefinition that serves all known use-cases. This CRD should follow the standard alpha -> beta -> GA feature process.
+
+Once the CRD API is defined we can move on with more delicate discussions about technology, organization and processes.
+
+### Goals
+
+- Define a alpha CRD
+- Fully document the Spec and use-cases
+
+### Non-Goals
+
+This KEP proposes the CRD Spec and documents the use-cases, not the choice of technology or migration path towards implementing the CRD.
+
+## Terminology
+
+* Kubernetes External Secrets `KES`: A Application that runs a control loop which syncs secrets
+* KES `instance`: A single entity that runs a control loop.
+* ExternalSecret `ES`: A CustomResource that declares which secrets should be synced
+* Store: Is a **source** for secrets. The Store is external to KES. It can be a hosted service like Alibabacloud SecretsManager, AWS SystemsManager, Azure KeyVault...
+* Frontend: A **sink** for the synced secrets. Usually a `ConfigMap` or `Secret`
+* Secret: credentials that act as a key to sensitive information
+
+## Use-Cases
+* one global KES instance that manages ES in **all namespaces**, which gives access to **all stores**, with ACL
+* multiple global KES instances, each manages access to a single or multiple stores (e.g.: shard by stage or team...)
+* one KES per namespace (a user manages his/her own KES instance)
+
+### User definitions
+* `operator :=` i manage one or multiple `KES` instances
+* `user :=` i only create `ES`, KES is managed by someone else
+
+### User Stories
+From that we can derive the following requirements or user-stories:
+1. AS a KES operator i want to run multiple KES instances per cluster (e.g. one KES instance per DEV/PROD)
+2. AS a KES operator or user i want to integrate **multiple stores** from a **single KES instance** (e.g. dev namespace has access only to dev secrets)
+3. AS a KES user i want to control the sink for the secrets (aka frontend: store secret as `kind=ConfigMap` or `kind=Secret`)
```

Comment:
Is storing the fetched ExternalSecrets into a ConfigMap something that should be solved? This seems likely to end up with secrets material stored in ConfigMaps potentially leading to other privilege escalations.

I do think there is a user story around needing to provide an application with a configuration, and having a secret be a field in the configuration. This could be solved with another solution like allowing templating within the ExternalSecret or referencing an existing ConfigMap with the configuration but without the secret.

Reply:
Config maps as a frontend seems funky, to me, in the context of secrets. Tho I could see there being use cases for it. There shouldn't really be any issue creating a configmap instead of a secret.

But perhaps in that case the actual resource should be different eg `ExternalConfig` instead of `ExternalSecret`? Not sure if there's added value in doing the separation, but I feel like it would be more explicit and allows more RBAC configuration

Reply:
I'm not sure I see the use case of either having the ExternalSecret create a ConfigMap, or having an ExternalConfig resource. Config should be part of application specification, within the helm chart or kubernetes manifests. This is safe to check in to git or other systems. ExternalSecrets, like ConfigMaps could also be safely stored in git and the secret is able to be fetched at deploy time,

Take something like Prometheus AlertManager config, which may need Slack/Pagerduty or other apikeys but is unable to source them from environment variables or other means (ref https://github.com/prometheus/alertmanager/issues/504), this still shouldn't have a frontend of a ConfigMap because it contains secrets.

I'd rather see ExternalSecret take the non-secret configuration and merge with the fetched external secrets, either directly or with a reference to another secret/configmap.

```yaml
spec:
  ... 
  data:
  - slack:
      key: /corp.org/slack
      property: api_url

# Secret is fetched and then merged and templated
  template:
    data:
      alertmanager.yaml: |
        receivers:
        - name: slack
          slack_configs:
          - channel: #alerts
            api_url: {{ .data.slack | quote }}
```

```yaml
spec:
  ... 
  data:
  - slack:
      key: /corp.org/slack
      property: api_url

# Secret is fetched and then merged and templated within the referenced configMap data
  templateFrom:
  - configMap:
      name: alertmanager
      items:
       - key: alertmanager.yaml
```

Reply:
I don't have a requirement for ConfigMaps. Lets remove it from this spec. If someone needs it later on we can add it.

Reply:
Thanks for the discussion! I'm soon gonna push commits that (hopefully) integrate your thoughts. Changelog:
* removed `ConfigMap`
* added `templateFrom` 

I'm resolving this conversation for now to tidy up the discussion. Feel free to un-resolve if there's something missing or wrong.